File: ttb_apps/ttb_apps/src/main/java/util/ros2udpProxy.cpp

#include <string>
#include <sstream>
#include <iostream>
#include <unistd.h>
#include "ros/ros.h"
#include <ros/transport_hints.h>
#include <stdio.h>


#include <SystemConfig.h>
#include <Configuration.h>
#include <exception>

#include <boost/asio.hpp>
#include <boost/thread.hpp> 

#include <sys/ioctl.h>
#include <net/if.h>
#include <netinet/in.h>
#include <arpa/inet.h>


import geometry_msgs.PoseWithCovarianceStamped;
import geometry_msgs.PoseArray;
import std_msgs.String;
import alica_ros_proxy.PlanTreeInfo;

using namespace supplementary;



using boost::asio::ip::udp;



std::string ownRosName;
udp::socket* insocket;
udp::endpoint otherEndPoint;
udp::endpoint destEndPoint;
boost::asio::ip::address multiCastAddress;
boost::asio::io_service io_service;
void handleUdpPacket(const boost::system::error_code& error,   std::size_t bytes_transferred);
void listenForPacket();



	private class OnRosPoseWithCovarianceStamped2852345798Listener implements MessageListener {
	@Override
public void onNewMessage(Object o) {
		PoseWithCovarianceStamped converted = (PoseWithCovarianceStamped) o;
		MessageSerializer<PoseWithCovarianceStamped> serializer = node.getMessageSerializationFactory().newMessageSerializer("geometry_msgs/PoseWithCovarianceStamped");
		ChannelBuffer buffer = ChannelBuffers.buffer(ByteOrder.LITTLE_ENDIAN,64000);
		serializer.serialize(converted,buffer);
		ByteBuffer idBuf = ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN).putInt((int) 2852345798l);
		ChannelBuffer finalBuf = ChannelBuffers.copiedBuffer(ByteOrder.LITTLE_ENDIAN, idBuf.array(), buffer.array());
		try {
			MulticastSocket socket = new MulticastSocket();
			socket.send(new DatagramPacket(finalBuf.array(),finalBuf.array().length,group,port));
			socket.close();
		} catch (IOException e) {
			System.err.println("Exception while sending UDP message:" + converted._TYPE + " Discarding message!");
		}

	}
	}

	private class OnRosPoseArray2644558886Listener implements MessageListener {
	@Override
public void onNewMessage(Object o) {
		PoseArray converted = (PoseArray) o;
		MessageSerializer<PoseArray> serializer = node.getMessageSerializationFactory().newMessageSerializer("geometry_msgs/PoseArray");
		ChannelBuffer buffer = ChannelBuffers.buffer(ByteOrder.LITTLE_ENDIAN,64000);
		serializer.serialize(converted,buffer);
		ByteBuffer idBuf = ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN).putInt((int) 2644558886l);
		ChannelBuffer finalBuf = ChannelBuffers.copiedBuffer(ByteOrder.LITTLE_ENDIAN, idBuf.array(), buffer.array());
		try {
			MulticastSocket socket = new MulticastSocket();
			socket.send(new DatagramPacket(finalBuf.array(),finalBuf.array().length,group,port));
			socket.close();
		} catch (IOException e) {
			System.err.println("Exception while sending UDP message:" + converted._TYPE + " Discarding message!");
		}

	}
	}

	private class OnRosString4022577436Listener implements MessageListener {
	@Override
public void onNewMessage(Object o) {
		String converted = (String) o;
		MessageSerializer<String> serializer = node.getMessageSerializationFactory().newMessageSerializer("std_msgs/String");
		ChannelBuffer buffer = ChannelBuffers.buffer(ByteOrder.LITTLE_ENDIAN,64000);
		serializer.serialize(converted,buffer);
		ByteBuffer idBuf = ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN).putInt((int) 4022577436l);
		ChannelBuffer finalBuf = ChannelBuffers.copiedBuffer(ByteOrder.LITTLE_ENDIAN, idBuf.array(), buffer.array());
		try {
			MulticastSocket socket = new MulticastSocket();
			socket.send(new DatagramPacket(finalBuf.array(),finalBuf.array().length,group,port));
			socket.close();
		} catch (IOException e) {
			System.err.println("Exception while sending UDP message:" + converted._TYPE + " Discarding message!");
		}

	}
	}

	private class OnRosPlanTreeInfo3767756765Listener implements MessageListener {
	@Override
public void onNewMessage(Object o) {
		PlanTreeInfo converted = (PlanTreeInfo) o;
		MessageSerializer<PlanTreeInfo> serializer = node.getMessageSerializationFactory().newMessageSerializer("alica_ros_proxy/PlanTreeInfo");
		ChannelBuffer buffer = ChannelBuffers.buffer(ByteOrder.LITTLE_ENDIAN,64000);
		serializer.serialize(converted,buffer);
		ByteBuffer idBuf = ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN).putInt((int) 3767756765l);
		ChannelBuffer finalBuf = ChannelBuffers.copiedBuffer(ByteOrder.LITTLE_ENDIAN, idBuf.array(), buffer.array());
		try {
			MulticastSocket socket = new MulticastSocket();
			socket.send(new DatagramPacket(finalBuf.array(),finalBuf.array().length,group,port));
			socket.close();
		} catch (IOException e) {
			System.err.println("Exception while sending UDP message:" + converted._TYPE + " Discarding message!");
		}

	}
	}


private Publisher<PoseWithCovarianceStamped> pub2852345798;
private Publisher<PoseArray> pub2644558886;
private Publisher<String> pub4022577436;
private Publisher<PlanTreeInfo> pub3767756765;

boost::array<char,64000> inBuffer;
void listenForPacket() {
	insocket->async_receive_from(boost::asio::buffer(inBuffer), otherEndPoint,
        boost::bind(&handleUdpPacket, boost::asio::placeholders::error,boost::asio::placeholders::bytes_transferred));
}
void handleUdpPacket(const boost::system::error_code& error,   std::size_t bytes_transferred) {
	//std::cout << "From "<<otherEndPoint.address() << std::endl;
	if (!error) { // && otherEndPoint.address() != localIP) {
		__uint32_t id = *((__uint32_t*)(inBuffer.data()));
		//std::cout << "Got packet"<<std::endl;
		try {	
			ros::serialization::IStream stream(((uint8_t*)inBuffer.data())+sizeof(__uint32_t),bytes_transferred-sizeof(__uint32_t));
			switch(id) {
if(id == 2852345798l) {
MessageDeserializer<PoseWithCovarianceStamped> deserializer = node.getMessageSerializationFactory().newMessageDeserializer(PoseWithCovarianceStamped._TYPE);
byte[] message = Arrays.copyOfRange(packet.getData(), Integer.SIZE / Byte.SIZE, packet.getData().length-4);
PoseWithCovarianceStamped m2852345798 = deserializer.deserialize(ChannelBuffers.copiedBuffer(ByteOrder.LITTLE_ENDIAN,message));
pub2852345798.publish(m2852345798);
}
else if(id == 2644558886l) {
MessageDeserializer<PoseArray> deserializer = node.getMessageSerializationFactory().newMessageDeserializer(PoseArray._TYPE);
byte[] message = Arrays.copyOfRange(packet.getData(), Integer.SIZE / Byte.SIZE, packet.getData().length-4);
PoseArray m2644558886 = deserializer.deserialize(ChannelBuffers.copiedBuffer(ByteOrder.LITTLE_ENDIAN,message));
pub2644558886.publish(m2644558886);
}
else if(id == 4022577436l) {
MessageDeserializer<String> deserializer = node.getMessageSerializationFactory().newMessageDeserializer(String._TYPE);
byte[] message = Arrays.copyOfRange(packet.getData(), Integer.SIZE / Byte.SIZE, packet.getData().length-4);
String m4022577436 = deserializer.deserialize(ChannelBuffers.copiedBuffer(ByteOrder.LITTLE_ENDIAN,message));
pub4022577436.publish(m4022577436);
}
else if(id == 3767756765l) {
MessageDeserializer<PlanTreeInfo> deserializer = node.getMessageSerializationFactory().newMessageDeserializer(PlanTreeInfo._TYPE);
byte[] message = Arrays.copyOfRange(packet.getData(), Integer.SIZE / Byte.SIZE, packet.getData().length-4);
PlanTreeInfo m3767756765 = deserializer.deserialize(ChannelBuffers.copiedBuffer(ByteOrder.LITTLE_ENDIAN,message));
pub3767756765.publish(m3767756765);
}
			
				default:
					std::cerr << "Cannot find Matching topic:" << id << std::endl;
			}
		}
		catch(std::exception& e) {
			ROS_ERROR_STREAM_THROTTLE(2,"Exception while receiving DDS message:"<<e.what()<< " Discarding message!");
		}
		
	}
	listenForPacket();
	return;
}
void run() {
	io_service.run();
}

int main (int argc, char *argv[])
{
	
	SystemConfig* sc = SystemConfig::getInstance();

	//Configuration *proxyconf = (*sc)["UdpProxy"];
udpConfig.load(getActivity().getResources().openRawResource(R.raw.ttb_apps));
	
	//std::string port = proxyconf->get<std::string>("UdpProxy","Port",NULL);
	
	std::string baddress = proxyconf->get<std::string>("UdpProxy","MulticastAddress",NULL);
	
	unsigned short port = (unsigned short)proxyconf->get<int>("UdpProxy","Port",NULL);
	
	//udp::resolver resolver(io_service);
    //udp::resolver::query query(udp::v4(), baddress, port);

	//bcastEndPoint = *resolver.resolve(query);
	
	
	
	
	
  
	//myEndPoint = udp::endpoint(udp::v4(),bcastEndPoint.port());
	
	multiCastAddress = boost::asio::ip::address::from_string(baddress);
	destEndPoint = udp::endpoint(multiCastAddress,port);
	
	std::cout<<"Opening to "<<multiCastAddress <<std::endl;
	
	
	
	insocket = new udp::socket(io_service,udp::endpoint(multiCastAddress,port));
	
	insocket->set_option(boost::asio::ip::multicast::enable_loopback(false));
	insocket->set_option(boost::asio::ip::multicast::join_group(multiCastAddress));
	listenForPacket();
	
	
 //   ros::init(argc, argv, "udpProxy");

    ros::NodeHandle n;
    ownRosName = ros::this_node::getName();//n.getNamespace();//n.resolveName("ddsProxy",true);
    
    std::cout << ownRosName << std::endl;
    
final Subscriber sub0 = connectedNode.newSubscriber("/amcl_pose", "geometry_msgs/PoseWithCovarianceStamped");
sub0.addMessageListener(new OnRosPoseWithCovarianceStamped2852345798Listener());
final Subscriber sub1 = connectedNode.newSubscriber("/particlecloud", "geometry_msgs/PoseArray");
sub1.addMessageListener(new OnRosPoseArray2644558886Listener());
final Subscriber sub2 = connectedNode.newSubscriber("/mmTalker", "std_msgs/String");
sub2.addMessageListener(new OnRosString4022577436Listener());
final Subscriber sub3 = connectedNode.newSubscriber("/AlicaEngine/PlanTreeInfo", "alica_ros_proxy/PlanTreeInfo");
sub3.addMessageListener(new OnRosPlanTreeInfo3767756765Listener());
	
pub2852345798 = connectedNode.newPublisher("/amcl_pose", "geometry_msgs/PoseWithCovarianceStamped");
pub2644558886 = connectedNode.newPublisher("/particlecloud", "geometry_msgs/PoseArray");
pub4022577436 = connectedNode.newPublisher("/mmTalker", "std_msgs/String");
pub3767756765 = connectedNode.newPublisher("/AlicaEngine/PlanTreeInfo", "alica_ros_proxy/PlanTreeInfo");
	
	boost::thread iothread(run);
    
    //ros::spin();
	ros::AsyncSpinner *spinner;
	
	spinner = new ros::AsyncSpinner(4);
	spinner->start();
	
	std::cout << "Ros2Udp Proxy running..." <<std::endl;
    while(n.ok()) {
	    usleep(300000);
    }
    io_service.stop();
    iothread.join();
    return 0;
}

File: ttb_apps/marauders_map/src/main/java/de/uni_kassel/vs/cn/ttb_apps/marauders_map/util/ROS2UDPProxy.java
package de.uni_kassel.vs.cn.ttb_apps.marauders_map.util;

import android.app.Activity;

import com.github.ros_java.marauders_map.R;
import org.jboss.netty.buffer.ChannelBuffer;
import org.jboss.netty.buffer.ChannelBuffers;
import org.ros.concurrent.CancellableLoop;
import org.ros.message.MessageDeserializer;
import org.ros.message.MessageListener;
import org.ros.message.MessageSerializer;
import org.ros.namespace.GraphName;
import org.ros.node.ConnectedNode;
import org.ros.node.Node;
import org.ros.node.NodeMain;
import org.ros.node.topic.Publisher;
import org.ros.node.topic.Subscriber;
import java.io.IOException;
import java.net.DatagramPacket;
import java.net.InetAddress;
import java.net.MulticastSocket;
import java.util.Arrays;
import java.nio.ByteBuffer;
import java.util.Properties;
import java.nio.ByteOrder;
import java.net.NetworkInterface;
import java.net.SocketException;
import java.util.Enumeration;
import java.net.Inet4Address;


import de.uni_kassel.vs.cn.ttb_apps.marauders_map.model.Root;
import ttb_msgs.GoalWrapped;
import ttb_msgs.InitialPoseWrapped;
import alica_ros_proxy.PlanTreeInfo;
import ttb_msgs.AMCLPoseWrapped;

/**
 * Created by marci on 22.10.15.
 */
public class ROS2UDPProxy implements NodeMain {

	private Activity activity;

	public java.lang.String ownRosName;

	private ConnectedNode node;

	private MulticastSocket udpSocket;

	private InetAddress group;

	private int port;

	private InetAddress localhost;

	public InetAddress getGroup() {
		return group;
	}

	public int getPort() {
		return port;
	}

	private class OnRosGoalWrapped3867808201Listener implements MessageListener {
		@Override
		public void onNewMessage(Object o) {
			GoalWrapped converted = (GoalWrapped) o;
			MessageSerializer<GoalWrapped> serializer = node.getMessageSerializationFactory().newMessageSerializer("ttb_msgs/GoalWrapped");
			ChannelBuffer buffer = ChannelBuffers.buffer(ByteOrder.LITTLE_ENDIAN,64000);
			serializer.serialize(converted,buffer);
			ByteBuffer idBuf = ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN).putInt((int) 3867808201l);
			ChannelBuffer finalBuf = ChannelBuffers.copiedBuffer(ByteOrder.LITTLE_ENDIAN, idBuf.array(), buffer.array());
			try {
				MulticastSocket socket = new MulticastSocket();
				socket.send(new DatagramPacket(finalBuf.array(),finalBuf.array().length,group,port));
				socket.close();
			} catch (IOException e) {
				System.err.println("Exception while sending UDP message:" + converted._TYPE + " Discarding message!");
			}

		}
	}

	{
		Root.topicHashmap.put("/wrapped/move_base_simple/goal", 3867808201l);}	private class OnRosInitialPoseWrapped1388158846Listener implements MessageListener {
		@Override
		public void onNewMessage(Object o) {
			InitialPoseWrapped converted = (InitialPoseWrapped) o;
			MessageSerializer<InitialPoseWrapped> serializer = node.getMessageSerializationFactory().newMessageSerializer("ttb_msgs/InitialPoseWrapped");
			ChannelBuffer buffer = ChannelBuffers.buffer(ByteOrder.LITTLE_ENDIAN,64000);
			serializer.serialize(converted,buffer);
			ByteBuffer idBuf = ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN).putInt((int) 1388158846l);
			ChannelBuffer finalBuf = ChannelBuffers.copiedBuffer(ByteOrder.LITTLE_ENDIAN, idBuf.array(), buffer.array());
			try {
				MulticastSocket socket = new MulticastSocket();
				socket.send(new DatagramPacket(finalBuf.array(),finalBuf.array().length,group,port));
				socket.close();
			} catch (IOException e) {
				System.err.println("Exception while sending UDP message:" + converted._TYPE + " Discarding message!");
			}

		}
	}

	{Root.topicHashmap.put("/wrapped/initialpose", 1388158846l);}
	private Publisher<GoalWrapped> pub3867808201;
	private Publisher<InitialPoseWrapped> pub1388158846;
	private Publisher<PlanTreeInfo> pub3767756765;
	private Publisher<AMCLPoseWrapped> pub2611391888;

	@Override
	public GraphName getDefaultNodeName() {
		return GraphName.of("MaraudersMap/Proxy");
	}

	@Override
	public void onStart(ConnectedNode connectedNode) {
		Properties udpConfig = new Properties();
		try {
			udpConfig.load(getActivity().getResources().openRawResource(R.raw.marauders_map));
			java.lang.String multiCastAdress = udpConfig.getProperty("MulticastAddress");
			port = Integer.parseInt(udpConfig.getProperty("Port"));
			udpSocket = new MulticastSocket(port);
			group = InetAddress.getByName(multiCastAdress);
			udpSocket.joinGroup(group);
			udpSocket.setLoopbackMode(false);
			node = connectedNode;

			final Subscriber sub0 = connectedNode.newSubscriber("/wrapped/move_base_simple/goal", "ttb_msgs/GoalWrapped");
			sub0.addMessageListener(new OnRosGoalWrapped3867808201Listener());
			final Subscriber sub1 = connectedNode.newSubscriber("/wrapped/initialpose", "ttb_msgs/InitialPoseWrapped");
			sub1.addMessageListener(new OnRosInitialPoseWrapped1388158846Listener());

			pub3867808201 = connectedNode.newPublisher("/wrapped/move_base_simple/goal", "ttb_msgs/GoalWrapped");
			pub1388158846 = connectedNode.newPublisher("/wrapped/initialpose", "ttb_msgs/InitialPoseWrapped");
			pub3767756765 = connectedNode.newPublisher("/AlicaEngine/PlanTreeInfo", "alica_ros_proxy/PlanTreeInfo");
			pub2611391888 = connectedNode.newPublisher("/wrapped/amcl_pose", "ttb_msgs/AMCLPoseWrapped");

			listenForPacket(udpSocket);
			ownRosName = connectedNode.getName().toString();

		} catch (IOException e) {
			e.printStackTrace();
		}
	}

	void listenForPacket(final MulticastSocket socket) {
		byte[] buffer = new byte[64000];

		try {
			Enumeration<NetworkInterface> list = NetworkInterface.getNetworkInterfaces();

			while(list.hasMoreElements() && localhost == null) {
				NetworkInterface networkInterface = list.nextElement();
				if(networkInterface.getName().contains("wlan")) {
					Enumeration<InetAddress> list2 = networkInterface.getInetAddresses();

					do {
						InetAddress address = list2.nextElement();
						if(address instanceof Inet4Address) {
							localhost = address;
						}
					} while (list2.hasMoreElements() && localhost == null);
				}
			}
		} catch (SocketException e) {
			e.printStackTrace();
		}

		final DatagramPacket packet = new DatagramPacket(buffer,buffer.length);
		node.executeCancellableLoop(new CancellableLoop() {
			@Override
			protected void loop() throws InterruptedException {
				try {
					socket.receive(packet);
					handleUdpPacket(packet);
				} catch (IOException e) {
					e.printStackTrace();
				}
			}
		});
	}

	@Override
	public void onShutdown(Node node) {

	}

	@Override
	public void onShutdownComplete(Node node) {

	}

	@Override
	public void onError(Node node, Throwable throwable) {
	}

	public Activity getActivity() {
		return activity;
	}

	public void setActivity(Activity activity) {
		this.activity = activity;
	}

	void handleUdpPacket(DatagramPacket packet) {
		if(packet.getAddress().equals(localhost)) {
			return;
		}
		long id = ChannelBuffers.copiedBuffer(ByteOrder.LITTLE_ENDIAN,Arrays.copyOfRange(packet.getData(), 0, 4)).getUnsignedInt(0);
		if(id == 3767756765l) {
			MessageDeserializer<PlanTreeInfo> deserializer = node.getMessageSerializationFactory().newMessageDeserializer(PlanTreeInfo._TYPE);
			byte[] message = Arrays.copyOfRange(packet.getData(), Integer.SIZE / Byte.SIZE, packet.getData().length-4);
			PlanTreeInfo m3767756765 = deserializer.deserialize(ChannelBuffers.copiedBuffer(ByteOrder.LITTLE_ENDIAN,message));
			pub3767756765.publish(m3767756765);
		}
		else if(id == 2611391888l) {
			MessageDeserializer<AMCLPoseWrapped> deserializer = node.getMessageSerializationFactory().newMessageDeserializer(AMCLPoseWrapped._TYPE);
			byte[] message = Arrays.copyOfRange(packet.getData(), Integer.SIZE / Byte.SIZE, packet.getData().length-4);
			AMCLPoseWrapped m2611391888 = deserializer.deserialize(ChannelBuffers.copiedBuffer(ByteOrder.LITTLE_ENDIAN,message));
			pub2611391888.publish(m2611391888);
		}
		else {
			System.err.println("Cannot find Matching topic:" + id);
		}
	}
}

